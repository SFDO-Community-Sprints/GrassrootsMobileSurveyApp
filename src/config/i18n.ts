import * as Localization from 'expo-localization';
import i18n from 'i18n-js';

import ne from './locale/ne_IN.json';
import en from './locale/en_US.json';

import { DEFAULT_SF_LANGUAGE } from '../constants';
import { toShortLocale } from '../utility';

//expo-localization gives the result in a language-region format
const deviceLocale: string =
  Localization.locale.indexOf('-') !== 0 ? Localization.locale.split('-')[0] : Localization.locale;

i18n.defaultLocale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.locale = deviceLocale === 'ne' ? deviceLocale : toShortLocale(DEFAULT_SF_LANGUAGE.code);
i18n.fallbacks = true;
i18n.translations = { en, ne };

export default i18n;
