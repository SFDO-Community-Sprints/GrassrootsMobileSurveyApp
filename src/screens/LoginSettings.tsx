import React, { useState, useEffect, useContext } from 'react';
import { Text, View, StyleSheet, ImageBackground } from 'react-native';
import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { Icon, Input, Button } from 'react-native-elements';
import * as SecureStore from 'expo-secure-store';

import {
  BACKGROUND_IMAGE_SOURCE,
  BACKGROUND_STYLE,
  BACKGROUND_IMAGE_STYLE,
  APP_FONTS,
  APP_THEME,
  SECURE_STORE_KEYS,
} from '../constants';
import LocalizationContext from '../context/localizationContext';
import { notifySuccess } from '../utility/notification';

export default function LoginSettings({ navigation }) {
  const [herokuAppName, setHerokuAppName] = useState('');

  const { t } = useContext(LocalizationContext);

  useEffect(() => {
    const prepare = async () => {
      const storedHerokuAppName = (await SecureStore.getItemAsync(SECURE_STORE_KEYS.HEROKU_APP_NAME)) || '';
      setHerokuAppName(storedHerokuAppName);
    };
    prepare();
  }, []);

  /**
   * @description Save Heroku app name.
   */
  const save = async () => {
    const trimmed = herokuAppName.trim();
    setHerokuAppName(trimmed);
    await SecureStore.setItemAsync(SECURE_STORE_KEYS.HEROKU_APP_NAME, trimmed);
    notifySuccess('Successfully saved!');
  };

  return (
    <ImageBackground source={BACKGROUND_IMAGE_SOURCE} style={BACKGROUND_STYLE} imageStyle={BACKGROUND_IMAGE_STYLE}>
      <KeyboardAwareScrollView style={styles.flex1}>
        <View style={styles.inputBoxesView}>
          <Input
            onChangeText={value => setHerokuAppName(value)}
            value={herokuAppName}
            label={t('HEROKU_APP_NAME')}
            labelStyle={{ fontFamily: APP_FONTS.FONT_REGULAR }}
            inputStyle={{ fontFamily: APP_FONTS.FONT_REGULAR }}
            placeholder="aaa-bbb-12345"
            keyboardType="url"
            accessibilityLabel="app-name-input"
          />
          <View style={styles.descriptionStyle}>
            <Icon name="information" type="material-community" color={APP_THEME.APP_DARK_FONT_COLOR} />
            <Text style={{ fontFamily: APP_FONTS.FONT_REGULAR, color: APP_THEME.APP_DARK_FONT_COLOR, paddingLeft: 5 }}>
              {t('HEROKU_APP_NAME_DESCRIPTION')}
            </Text>
          </View>
          <View style={styles.saveButtonContainer}>
            <Button
              title={t('SAVE')}
              onPress={() => save()}
              buttonStyle={styles.saveButtonBackground}
              titleStyle={styles.saveButtonText}
              disabled={!herokuAppName}
            />
          </View>
        </View>
      </KeyboardAwareScrollView>
    </ImageBackground>
  );
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    alignItems: 'center',
    justifyContent: 'flex-end',
    marginTop: 20,
  },
  flex1: {
    flex: 1,
  },
  inputBoxesView: {
    width: '90%',
    paddingTop: 20,
    alignItems: 'center',
    justifyContent: 'center',
    alignSelf: 'center',
  },
  descriptionStyle: {
    paddingLeft: 10,
    paddingRight: 10,
    flexDirection: 'row',
  },
  saveButtonContainer: { width: '40%', alignSelf: 'center', paddingTop: 20 },
  saveButtonBackground: {
    backgroundColor: APP_THEME.APP_BASE_FONT_COLOR,
    borderRadius: 5,
  },
  saveButtonText: {
    fontSize: 16,
    fontFamily: APP_FONTS.FONT_REGULAR,
  },
});
