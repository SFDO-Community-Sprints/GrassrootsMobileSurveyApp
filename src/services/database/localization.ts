import { ASYNC_STORAGE_KEYS } from '../../constants';
import { SQLiteFieldTypeMapping } from '../../types/sqlite';
import { getAvailableLanguages } from '../salesforce/metadata';
import { prepareTable } from './database';

export const prepareLocalizationTable = async () => {
  const fieldTypeMappings: Array<SQLiteFieldTypeMapping> = [
    { name: 'Type__c', type: 'text' },
    { name: 'Locale__c', type: 'text' },
    { name: 'OriginalName__c', type: 'text' },
    { name: 'TranslatedLabel__c', type: 'text' },
  ];
  await prepareTable('Localization', fieldTypeMappings, undefined);
};

export const storeAvailableLanguages = async () => {
  const availableLanguages = await getAvailableLanguages();
  storage.save({ key: ASYNC_STORAGE_KEYS.LANGUAGES, data: availableLanguages });
};
