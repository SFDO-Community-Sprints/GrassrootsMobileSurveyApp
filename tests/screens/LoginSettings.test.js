import React from 'react';
import { render, fireEvent, waitFor } from '@testing-library/react-native';

import i18n from '../../src/config/i18n';
import LocalizationContext from '../../src/context/localizationContext';
import LoginSettings from '../../src/screens/LoginSettings';
// local test utils
import { translate } from '../localization';

jest.mock('react-native-keyboard-aware-scroll-view', () => {
  const KeyboardAwareScrollView = ({ children }) => children;
  return { KeyboardAwareScrollView };
});

describe('login settings screen', () => {
  const locale = i18n.locale;
  const localizationContext = {
    t: jest.fn().mockImplementation(key => translate(key)),
    locale,
    setLocale: jest.fn(),
  };

  it('initial render', () => {
    const screen = render(
      <LocalizationContext.Provider value={localizationContext}>
        <LoginSettings />
      </LocalizationContext.Provider>
    );

    // expect one input field
    expect(screen.getByLabelText('app-name-input')).toBeTruthy();

    // save button disabled when blank
    const saveButton = screen.getByA11yRole('button');
    expect(saveButton).toBeDisabled();
  });

  it('save', async () => {
    const screen = render(
      <LocalizationContext.Provider value={localizationContext}>
        <LoginSettings />
      </LocalizationContext.Provider>
    );

    const appNameInput = screen.getByLabelText('app-name-input');
    fireEvent.changeText(appNameInput, 'test-app-123');
    const saveButton = screen.getByA11yRole('button');
    fireEvent.press(saveButton);
    expect(screen.toJSON()).toMatchSnapshot(); // value='test-app-123'
    await waitFor(() => {
      Promise.resolve();
    });
  });

  it('save with trim', async () => {
    const screen = render(
      <LocalizationContext.Provider value={localizationContext}>
        <LoginSettings />
      </LocalizationContext.Provider>
    );

    const appNameInput = screen.getByLabelText('app-name-input');
    fireEvent.changeText(appNameInput, '  test-app-123  ');

    const saveButton = screen.getByA11yRole('button');
    fireEvent.press(saveButton);

    expect(screen.toJSON()).toMatchSnapshot(); // value='test-app-123'
    await waitFor(() => {
      Promise.resolve();
    });
  });
});
