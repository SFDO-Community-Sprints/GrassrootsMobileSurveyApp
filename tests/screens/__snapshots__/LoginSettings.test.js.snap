// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`login settings screen save 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        Object {
          "resizeMode": "contain",
          "top": 857.7142857142858,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "justifyContent": "center",
        "paddingTop": 20,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#86939e",
            "fontFamily": "SalesforceSans-Regular",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Heroku App Name
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          accessibilityLabel="app-name-input"
          allowFontScaling={true}
          editable={true}
          keyboardType="url"
          onChangeText={[Function]}
          placeholder="aaa-bbb-12345"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          style={
            Object {
              "alignSelf": "center",
              "color": "#242424",
              "flex": 1,
              "fontFamily": "SalesforceSans-Regular",
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value="test-app-123"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <View
        color="#2b2826"
        name="information"
        type="material-community"
      />
      <Text
        style={
          Object {
            "color": "#2b2826",
            "fontFamily": "SalesforceSans-Regular",
            "paddingLeft": 5,
          }
        }
      >
        Enter your Heroku app name. For example, if the URL of your Heroku app is https://aaa-bbb-12345.herokuapp.com, enter aaa-bbb-12345. Ask your administrator for details.
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "paddingTop": 20,
          "width": "40%",
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 5,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#01579b",
                "borderColor": "#2089dc",
                "borderRadius": 5,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontFamily": "SalesforceSans-Regular",
                  "fontSize": 16,
                  "paddingBottom": 1,
                  "paddingTop": 2,
                  "textAlign": "center",
                }
              }
            >
              Save
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`login settings screen save with trim 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Object {
      "backgroundColor": "white",
      "flex": 1,
    }
  }
>
  <Image
    source={1}
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        Object {
          "resizeMode": "contain",
          "top": 857.7142857142858,
        },
      ]
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "alignSelf": "center",
        "justifyContent": "center",
        "paddingTop": 20,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "paddingHorizontal": 10,
          "width": "100%",
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#86939e",
            "fontFamily": "SalesforceSans-Regular",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        Heroku App Name
      </Text>
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 1,
            "borderColor": "#86939e",
            "flexDirection": "row",
            "transform": Array [
              Object {
                "translateX": 0,
              },
            ],
          }
        }
      >
        <TextInput
          accessibilityLabel="app-name-input"
          allowFontScaling={true}
          editable={true}
          keyboardType="url"
          onChangeText={[Function]}
          placeholder="aaa-bbb-12345"
          placeholderTextColor="#86939e"
          rejectResponderTermination={true}
          style={
            Object {
              "alignSelf": "center",
              "color": "#242424",
              "flex": 1,
              "fontFamily": "SalesforceSans-Regular",
              "fontSize": 18,
              "minHeight": 40,
            }
          }
          testID="RNE__Input__text-input"
          underlineColorAndroid="transparent"
          value="test-app-123"
        />
      </View>
      <Text
        style={
          Object {
            "color": "#ff190c",
            "fontSize": 12,
            "margin": 5,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "flexDirection": "row",
          "paddingLeft": 10,
          "paddingRight": 10,
        }
      }
    >
      <View
        color="#2b2826"
        name="information"
        type="material-community"
      />
      <Text
        style={
          Object {
            "color": "#2b2826",
            "fontFamily": "SalesforceSans-Regular",
            "paddingLeft": 5,
          }
        }
      >
        Enter your Heroku app name. For example, if the URL of your Heroku app is https://aaa-bbb-12345.herokuapp.com, enter aaa-bbb-12345. Ask your administrator for details.
      </Text>
    </View>
    <View
      style={
        Object {
          "alignSelf": "center",
          "paddingTop": 20,
          "width": "40%",
        }
      }
    >
      <View
        style={
          Object {
            "borderRadius": 5,
            "overflow": "hidden",
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#01579b",
                "borderColor": "#2089dc",
                "borderRadius": 5,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "white",
                  "fontFamily": "SalesforceSans-Regular",
                  "fontSize": 16,
                  "paddingBottom": 1,
                  "paddingTop": 2,
                  "textAlign": "center",
                }
              }
            >
              Save
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
